{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "___________________________________________________________________________________________________________________________"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "___________________________________________________________________________________________________________________________"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Insert the correct project ID number as string"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#provide project number ID as string (PO number in string format)\n",
    "project_id = 'project_test'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "___________________________________________________________________________________________________________________________"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "___________________________________________________________________________________________________________________________"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#import for retrieving predicted gRNAs\n",
    "from pybiomart import Server\n",
    "import pandas as pd\n",
    "import urllib.request\n",
    "import os\n",
    "# import nest_asyncio\n",
    "# nest_asyncio.apply()\n",
    "from selenium import webdriver\n",
    "import time\n",
    "pd.set_option('display.max_rows', None)\n",
    "pd.options.display.max_columns = None\n",
    "#imports for primer design\n",
    "from Bio.Seq import Seq\n",
    "import primer3\n",
    "import numpy as np\n",
    "import requests\n",
    "import sys\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# define functions\n",
    "def predict_gRNA(df_genes):\n",
    "    \n",
    "    #uses globally declared dataframe, declared prior to function calls\n",
    "    global gRNA_predicted\n",
    "    \n",
    "    for entry in range(len(df_genes)):\n",
    "        gene_ens_id = df_genes[\"GENE ID\"].iloc[entry]\n",
    "\n",
    "        #parse dataset for gene_id\n",
    "        data_gene = dataset.query(attributes=['ensembl_gene_id', 'external_gene_name'],\n",
    "                      filters={'link_ensembl_gene_id': gene_ens_id})\n",
    "\n",
    "        #if gene_id found and unique, construct Synthego url\n",
    "        if len(data_gene.index) == 1:\n",
    "            #print(data_gene)\n",
    "            gene_id = data_gene.iat[0,0]\n",
    "            gene_symbol = data_gene.iat[0,1]\n",
    "            url = 'https://design.synthego.com/#/design/results?genome=homo_sapiens_gencode_26_primary&nuclease=cas9&gene_id='+gene_id+'&symbol='+gene_symbol\n",
    "            # specify the url\n",
    "            urlpage = url\n",
    "            #print(urlpage)\n",
    "            #suppress opening Firefox browser window\n",
    "            os.environ['MOZ_HEADLESS'] = '1'\n",
    "            # run firefox webdriver from executable path of your choice\n",
    "            driver = webdriver.Firefox(executable_path = 'D:\\Alex\\geckodriver-v0.31.0-win64\\geckodriver.exe')\n",
    "            # get web page\n",
    "            driver.get(urlpage)\n",
    "            #30 seconds wait to have page fully loaded - might need to increase time\n",
    "            time.sleep(30)\n",
    "            results = driver.find_elements(\"xpath\", \"//*[@class='ng-binding']\")\n",
    "            #print('Number of results', len(results))\n",
    "            #Store predictions\n",
    "            gRNAs_synth = []\n",
    "            # loop over results\n",
    "            for result in results:\n",
    "                #print(result.text)\n",
    "                gRNAs_synth.append(result.text)\n",
    "            #extract minimal information for predicted gRNAs\n",
    "            gRNAs_synth_min = gRNAs_synth[3:9]\n",
    "            gRNAs_synth_min_df = pd.DataFrame(gRNAs_synth_min[2:], index =[gene_symbol+'-gRNA1', gene_symbol+'-gRNA2', gene_symbol+'-gRNA3', gene_symbol+'-gRNA4'],\n",
    "                                                 columns =['gRNA sequence'])\n",
    "            #add column(s) with ENSG number/gene name to df\n",
    "            gRNAs_synth_min_df['GENE ID'] = gene_id\n",
    "            gRNAs_synth_min_df['GENE NAME'] = gene_symbol\n",
    "            #print(gRNAs_synth_min_df)\n",
    "            gRNA_predicted = pd.concat([gRNA_predicted, gRNAs_synth_min_df])\n",
    "\n",
    "        else:\n",
    "            raise SystemExit(\"Stop right there - ambiguous gene name!\")   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#connect to Biomart Server\n",
    "server = Server(host='http://www.ensembl.org')\n",
    "\n",
    "# use #dataset.list_filters() to see available filters for list_filters method\n",
    "#dataset.list_filters()\n",
    "\n",
    "#generate dataset (homo sapiens)\n",
    "dataset = (server.marts['ENSEMBL_MART_ENSEMBL']\n",
    "                 .datasets['hsapiens_gene_ensembl'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE ID</th>\n",
       "      <th>GENE NAME</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000010404</td>\n",
       "      <td>IDS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000150676</td>\n",
       "      <td>CCDC83</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000099308</td>\n",
       "      <td>MAST3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000142676</td>\n",
       "      <td>RPL11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000180776</td>\n",
       "      <td>ZDHHC20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSG00000108848</td>\n",
       "      <td>LUC7L3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSG00000103811</td>\n",
       "      <td>CTSH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSG00000101473</td>\n",
       "      <td>ACOT8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENSG00000115414</td>\n",
       "      <td>FN1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENSG00000156886</td>\n",
       "      <td>ITGAD</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           GENE ID GENE NAME\n",
       "0  ENSG00000010404       IDS\n",
       "1  ENSG00000150676    CCDC83\n",
       "2  ENSG00000099308     MAST3\n",
       "3  ENSG00000142676     RPL11\n",
       "4  ENSG00000180776   ZDHHC20\n",
       "5  ENSG00000108848    LUC7L3\n",
       "6  ENSG00000103811      CTSH\n",
       "7  ENSG00000101473     ACOT8\n",
       "8  ENSG00000115414       FN1\n",
       "9  ENSG00000156886     ITGAD"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#read gene ID numbers from csv file - gene ID in form of ENSG00000141510\n",
    "df_genes_id = pd.read_csv(project_id + '.csv')\n",
    "df_genes_id.columns = df_genes_id.columns.str.upper()\n",
    "df_genes_id.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#loop through gene IDs, retrieve gene name (to double check with client data) \n",
    "#predict and scrape gRNAs\n",
    "# create df for storage of predicted gRNAs\n",
    "gRNA_predicted = pd.DataFrame(columns=['gRNA sequence','Gene ID', 'Gene Name'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "#call function to retrieve predicted gRNAs\n",
    "predict_gRNA(df_genes_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Empty DataFrame\n",
      "Columns: [gRNA sequence, Gene ID, Gene Name, GENE ID, GENE NAME]\n",
      "Index: []\n"
     ]
    }
   ],
   "source": [
    "#check if all genes have predicted gRNAs - if not, run predicted_gRNA function until all genes have predicted gRNAs)\n",
    "while len(gRNA_predicted[gRNA_predicted['gRNA sequence'].isnull()]) > 0:\n",
    "    print(gRNA_predicted[gRNA_predicted['gRNA sequence'].isnull()])\n",
    "    nan_rows = gRNA_predicted[gRNA_predicted['gRNA sequence'].isnull()]\n",
    "    nan_rows = nan_rows.drop_duplicates(subset=None, keep='first', inplace=False, ignore_index='False')\n",
    "    nan_rows = nan_rows[['GENE ID', 'GENE NAME']]\n",
    "    nan_rows.columns = nan_rows.columns.str.upper()\n",
    "    predict_gRNA(nan_rows)\n",
    "    \n",
    "print(gRNA_predicted[gRNA_predicted['gRNA sequence'].isnull()])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gRNA sequence</th>\n",
       "      <th>GENE ID</th>\n",
       "      <th>GENE NAME</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>IDS-gRNA1</th>\n",
       "      <td>GAACGUUCUUCUCAUCAUCG</td>\n",
       "      <td>ENSG00000010404</td>\n",
       "      <td>IDS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IDS-gRNA2</th>\n",
       "      <td>CCCCAUAACAGCCCAGGGAG</td>\n",
       "      <td>ENSG00000010404</td>\n",
       "      <td>IDS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IDS-gRNA3</th>\n",
       "      <td>GAUGACCUGCGCCCCUCCCU</td>\n",
       "      <td>ENSG00000010404</td>\n",
       "      <td>IDS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IDS-gRNA4</th>\n",
       "      <td>CUUAUCCCCAUAACAGCCCA</td>\n",
       "      <td>ENSG00000010404</td>\n",
       "      <td>IDS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCDC83-gRNA1</th>\n",
       "      <td>AAAUCUGUUCUUCUUUUAAG</td>\n",
       "      <td>ENSG00000150676</td>\n",
       "      <td>CCDC83</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCDC83-gRNA2</th>\n",
       "      <td>GCUUAAAAGAAGAACAGAUU</td>\n",
       "      <td>ENSG00000150676</td>\n",
       "      <td>CCDC83</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCDC83-gRNA3</th>\n",
       "      <td>AAGAACAGAUUUGGCACAUA</td>\n",
       "      <td>ENSG00000150676</td>\n",
       "      <td>CCDC83</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCDC83-gRNA4</th>\n",
       "      <td>GCACAUACGGCAUCUACUAA</td>\n",
       "      <td>ENSG00000150676</td>\n",
       "      <td>CCDC83</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MAST3-gRNA1</th>\n",
       "      <td>CCCGGGAGGAUGAGCUUGAC</td>\n",
       "      <td>ENSG00000099308</td>\n",
       "      <td>MAST3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MAST3-gRNA2</th>\n",
       "      <td>ACCAGUCAAGCUCAUCCUCC</td>\n",
       "      <td>ENSG00000099308</td>\n",
       "      <td>MAST3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     gRNA sequence          GENE ID GENE NAME\n",
       "IDS-gRNA1     GAACGUUCUUCUCAUCAUCG  ENSG00000010404       IDS\n",
       "IDS-gRNA2     CCCCAUAACAGCCCAGGGAG  ENSG00000010404       IDS\n",
       "IDS-gRNA3     GAUGACCUGCGCCCCUCCCU  ENSG00000010404       IDS\n",
       "IDS-gRNA4     CUUAUCCCCAUAACAGCCCA  ENSG00000010404       IDS\n",
       "CCDC83-gRNA1  AAAUCUGUUCUUCUUUUAAG  ENSG00000150676    CCDC83\n",
       "CCDC83-gRNA2  GCUUAAAAGAAGAACAGAUU  ENSG00000150676    CCDC83\n",
       "CCDC83-gRNA3  AAGAACAGAUUUGGCACAUA  ENSG00000150676    CCDC83\n",
       "CCDC83-gRNA4  GCACAUACGGCAUCUACUAA  ENSG00000150676    CCDC83\n",
       "MAST3-gRNA1   CCCGGGAGGAUGAGCUUGAC  ENSG00000099308     MAST3\n",
       "MAST3-gRNA2   ACCAGUCAAGCUCAUCCUCC  ENSG00000099308     MAST3"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#clean gRNA_predicted df\n",
    "gRNA_predicted = gRNA_predicted.drop(['Gene ID', 'Gene Name'], axis=1)\n",
    "#save predicted gRNAs as csv\n",
    "gRNA_predicted.to_csv(project_id + '_gRNA_predicted.csv') \n",
    "#sanity check df\n",
    "gRNA_predicted.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gRNA sequence</th>\n",
       "      <th>GENE ID</th>\n",
       "      <th>GENE NAME</th>\n",
       "      <th>gRNA_sequence_T</th>\n",
       "      <th>gRNA_sequence_T_forward_rel_to_gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>IDS-gRNA1</th>\n",
       "      <td>GAACGUUCUUCUCAUCAUCG</td>\n",
       "      <td>ENSG00000010404</td>\n",
       "      <td>IDS</td>\n",
       "      <td>GAACGTTCTTCTCATCATCG</td>\n",
       "      <td>GAACGTTCTTCTCATCATCG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IDS-gRNA2</th>\n",
       "      <td>CCCCAUAACAGCCCAGGGAG</td>\n",
       "      <td>ENSG00000010404</td>\n",
       "      <td>IDS</td>\n",
       "      <td>CCCCATAACAGCCCAGGGAG</td>\n",
       "      <td>CCCCATAACAGCCCAGGGAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IDS-gRNA3</th>\n",
       "      <td>GAUGACCUGCGCCCCUCCCU</td>\n",
       "      <td>ENSG00000010404</td>\n",
       "      <td>IDS</td>\n",
       "      <td>GATGACCTGCGCCCCTCCCT</td>\n",
       "      <td>GATGACCTGCGCCCCTCCCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IDS-gRNA4</th>\n",
       "      <td>CUUAUCCCCAUAACAGCCCA</td>\n",
       "      <td>ENSG00000010404</td>\n",
       "      <td>IDS</td>\n",
       "      <td>CTTATCCCCATAACAGCCCA</td>\n",
       "      <td>CTTATCCCCATAACAGCCCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCDC83-gRNA1</th>\n",
       "      <td>AAAUCUGUUCUUCUUUUAAG</td>\n",
       "      <td>ENSG00000150676</td>\n",
       "      <td>CCDC83</td>\n",
       "      <td>AAATCTGTTCTTCTTTTAAG</td>\n",
       "      <td>AAATCTGTTCTTCTTTTAAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCDC83-gRNA2</th>\n",
       "      <td>GCUUAAAAGAAGAACAGAUU</td>\n",
       "      <td>ENSG00000150676</td>\n",
       "      <td>CCDC83</td>\n",
       "      <td>GCTTAAAAGAAGAACAGATT</td>\n",
       "      <td>GCTTAAAAGAAGAACAGATT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCDC83-gRNA3</th>\n",
       "      <td>AAGAACAGAUUUGGCACAUA</td>\n",
       "      <td>ENSG00000150676</td>\n",
       "      <td>CCDC83</td>\n",
       "      <td>AAGAACAGATTTGGCACATA</td>\n",
       "      <td>AAGAACAGATTTGGCACATA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCDC83-gRNA4</th>\n",
       "      <td>GCACAUACGGCAUCUACUAA</td>\n",
       "      <td>ENSG00000150676</td>\n",
       "      <td>CCDC83</td>\n",
       "      <td>GCACATACGGCATCTACTAA</td>\n",
       "      <td>GCACATACGGCATCTACTAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MAST3-gRNA1</th>\n",
       "      <td>CCCGGGAGGAUGAGCUUGAC</td>\n",
       "      <td>ENSG00000099308</td>\n",
       "      <td>MAST3</td>\n",
       "      <td>CCCGGGAGGATGAGCTTGAC</td>\n",
       "      <td>CCCGGGAGGATGAGCTTGAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MAST3-gRNA2</th>\n",
       "      <td>ACCAGUCAAGCUCAUCCUCC</td>\n",
       "      <td>ENSG00000099308</td>\n",
       "      <td>MAST3</td>\n",
       "      <td>ACCAGTCAAGCTCATCCTCC</td>\n",
       "      <td>ACCAGTCAAGCTCATCCTCC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     gRNA sequence          GENE ID GENE NAME  \\\n",
       "IDS-gRNA1     GAACGUUCUUCUCAUCAUCG  ENSG00000010404       IDS   \n",
       "IDS-gRNA2     CCCCAUAACAGCCCAGGGAG  ENSG00000010404       IDS   \n",
       "IDS-gRNA3     GAUGACCUGCGCCCCUCCCU  ENSG00000010404       IDS   \n",
       "IDS-gRNA4     CUUAUCCCCAUAACAGCCCA  ENSG00000010404       IDS   \n",
       "CCDC83-gRNA1  AAAUCUGUUCUUCUUUUAAG  ENSG00000150676    CCDC83   \n",
       "CCDC83-gRNA2  GCUUAAAAGAAGAACAGAUU  ENSG00000150676    CCDC83   \n",
       "CCDC83-gRNA3  AAGAACAGAUUUGGCACAUA  ENSG00000150676    CCDC83   \n",
       "CCDC83-gRNA4  GCACAUACGGCAUCUACUAA  ENSG00000150676    CCDC83   \n",
       "MAST3-gRNA1   CCCGGGAGGAUGAGCUUGAC  ENSG00000099308     MAST3   \n",
       "MAST3-gRNA2   ACCAGUCAAGCUCAUCCUCC  ENSG00000099308     MAST3   \n",
       "\n",
       "                   gRNA_sequence_T gRNA_sequence_T_forward_rel_to_gene  \n",
       "IDS-gRNA1     GAACGTTCTTCTCATCATCG                GAACGTTCTTCTCATCATCG  \n",
       "IDS-gRNA2     CCCCATAACAGCCCAGGGAG                CCCCATAACAGCCCAGGGAG  \n",
       "IDS-gRNA3     GATGACCTGCGCCCCTCCCT                GATGACCTGCGCCCCTCCCT  \n",
       "IDS-gRNA4     CTTATCCCCATAACAGCCCA                CTTATCCCCATAACAGCCCA  \n",
       "CCDC83-gRNA1  AAATCTGTTCTTCTTTTAAG                AAATCTGTTCTTCTTTTAAG  \n",
       "CCDC83-gRNA2  GCTTAAAAGAAGAACAGATT                GCTTAAAAGAAGAACAGATT  \n",
       "CCDC83-gRNA3  AAGAACAGATTTGGCACATA                AAGAACAGATTTGGCACATA  \n",
       "CCDC83-gRNA4  GCACATACGGCATCTACTAA                GCACATACGGCATCTACTAA  \n",
       "MAST3-gRNA1   CCCGGGAGGATGAGCTTGAC                CCCGGGAGGATGAGCTTGAC  \n",
       "MAST3-gRNA2   ACCAGTCAAGCTCATCCTCC                ACCAGTCAAGCTCATCCTCC  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#replace U with T in gRNA sequences\n",
    "gRNA_predicted[\"gRNA_sequence_T\"] = gRNA_predicted[\"gRNA sequence\"].str.replace(\"U\",'T')\n",
    "#add column to indicate forward/reverse orientation of gRNA in relation to gene sequence\n",
    "gRNA_predicted[\"gRNA_sequence_T_forward_rel_to_gene\"] = gRNA_predicted[\"gRNA_sequence_T\"]\n",
    "\n",
    "#sanity check the dataframe\n",
    "gRNA_predicted.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "#REST API python3 Ensembl\n",
    "server = \"https://rest.ensembl.org\"\n",
    "#initiate dict to store gene sequences for all genes in gRNA_predicted\n",
    "gene_sequences = {}\n",
    "\n",
    "# get unique gene IDs from gRNA_predicted\n",
    "unique_gene_IDs = gRNA_predicted['GENE ID'].unique()\n",
    "\n",
    "for entry in unique_gene_IDs:\n",
    "    ext = \"/sequence/id/\" + entry +\"?\"\n",
    "    # retrieve plain text gene sequence\n",
    "    r = requests.get(server+ext, headers={ \"Content-Type\" : \"text/plain\"})\n",
    "    # sanity check object r\n",
    "    if not r.ok:\n",
    "        r.raise_for_status()\n",
    "        sys.exit()\n",
    "    gene_sequences[entry] = r.text \n",
    "#pd.DataFrame.from_dict(gene_sequences, orient='index')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[16810, -1, 16832, -1, -1, 27419, 27428, 27441, -1, 25434, 25435, -1, 855, 863, -1, -1, -1, 38207, -1, -1, -1, 21547, 21605, 21616, -1, 12150, 12153, 12167, 2107, 2113, -1, 2121, -1, 1245, 1273, 1274, 884, -1, -1, 908, -1, 994, -1, 1010, -1, 66046, -1, 66102, 971, -1, -1, -1, 491, -1, -1, 524]\n"
     ]
    }
   ],
   "source": [
    "#sanity check for orientation and existence of gRNA_sequence_T in gene sequences \n",
    "for_rev_orientation_gRNA = []\n",
    "\n",
    "#using .find to match gRNA sequence_T to gene sequence - if not found output will be -1.\n",
    "for row in range(len(gRNA_predicted)):\n",
    "        #get key from gRNA_predicted to retrieve value from gene_sequences dict\n",
    "        key = gRNA_predicted[\"GENE ID\"].iloc[row]\n",
    "        value = gene_sequences[key]\n",
    "        orientation = value.find(gRNA_predicted[\"gRNA_sequence_T\"].iloc[row])\n",
    "        for_rev_orientation_gRNA.append(orientation)\n",
    "print(for_rev_orientation_gRNA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "gRNA_predicted['gRNA reverse (-1 = True)'] = for_rev_orientation_gRNA\n",
    "#gRNA_predicted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "#In case of output -1 in for_rev_orientation_gRNA use reverse complement gRNA to search for match\n",
    "for entry in range(len(gRNA_predicted)):\n",
    "    if gRNA_predicted.iat[entry, 5] == -1:\n",
    "        gRNA_predicted[\"gRNA_sequence_T_forward_rel_to_gene\"].iloc[entry] = Seq(gRNA_predicted[\"gRNA_sequence_T\"].iloc[entry]).reverse_complement()\n",
    "        gRNA_predicted[\"gRNA_sequence_T_forward_rel_to_gene\"].iloc[entry] = str(gRNA_predicted[\"gRNA_sequence_T_forward_rel_to_gene\"].iloc[entry])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[16810, 16846, 16832, 16851, 27420, 27419, 27428, 27441, 25438, 25434, 25435, 25452, 855, 863, 893, 894, 38215, 38207, 38225, 38237, 21536, 21547, 21605, 21616, 12164, 12150, 12153, 12167, 2107, 2113, 2135, 2121, 1252, 1245, 1273, 1274, 884, 898, 905, 908, 1002, 994, 1012, 1010, 66060, 66046, 66081, 66102, 971, 1003, 1018, 1019, 491, 526, 532, 524]\n"
     ]
    }
   ],
   "source": [
    "#sanity check for orientation and existence of gRNA_sequence_T in gene sequences \n",
    "for_orientation_gRNA = []\n",
    "\n",
    "#using .find to match gRNA sequence_T to gene sequence - if not found output will be -1.\n",
    "#In case of output -1 use reverse complement gRNA to search for match\n",
    "for row in range(len(gRNA_predicted)):\n",
    "        #get key from gRNA_predicted to retrieve value from gene_sequences dict\n",
    "        key = gRNA_predicted[\"GENE ID\"].iloc[row]\n",
    "        value = gene_sequences[key]\n",
    "        orientation = value.find(gRNA_predicted[\"gRNA_sequence_T_forward_rel_to_gene\"].iloc[row])\n",
    "        for_orientation_gRNA.append(orientation)\n",
    "print(for_orientation_gRNA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "gRNA_predicted[\"Forward position gRNA\"] = for_orientation_gRNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Extract gRNA flanking sequences ~250 nt for primer design\n",
    "#dict to hold sequences\n",
    "sequences_for_primer_design = []\n",
    "\n",
    "for row in range(len(gRNA_predicted)):\n",
    "    beginning = gRNA_predicted['Forward position gRNA'].iloc[row] - 250\n",
    "    end = gRNA_predicted['Forward position gRNA'].iloc[row] + 250\n",
    "    key = gRNA_predicted[\"GENE ID\"].iloc[row]\n",
    "    value = gene_sequences[key]\n",
    "    seq_slice = value[beginning:end]\n",
    "    sequences_for_primer_design.append(seq_slice)\n",
    "#sanity check dict\n",
    "#sequences_for_primer_design"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gRNA sequence</th>\n",
       "      <th>GENE ID</th>\n",
       "      <th>GENE NAME</th>\n",
       "      <th>gRNA_sequence_T</th>\n",
       "      <th>gRNA_sequence_T_forward_rel_to_gene</th>\n",
       "      <th>gRNA reverse (-1 = True)</th>\n",
       "      <th>Forward position gRNA</th>\n",
       "      <th>Seq for primer design</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>IDS-gRNA1</th>\n",
       "      <td>GAACGUUCUUCUCAUCAUCG</td>\n",
       "      <td>ENSG00000010404</td>\n",
       "      <td>IDS</td>\n",
       "      <td>GAACGTTCTTCTCATCATCG</td>\n",
       "      <td>GAACGTTCTTCTCATCATCG</td>\n",
       "      <td>16810</td>\n",
       "      <td>16810</td>\n",
       "      <td>CACTCACTATCTCGCTTCCTCACCCAGGTTGGAGGCCATGACCAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IDS-gRNA2</th>\n",
       "      <td>CCCCAUAACAGCCCAGGGAG</td>\n",
       "      <td>ENSG00000010404</td>\n",
       "      <td>IDS</td>\n",
       "      <td>CCCCATAACAGCCCAGGGAG</td>\n",
       "      <td>CTCCCTGGGCTGTTATGGGG</td>\n",
       "      <td>-1</td>\n",
       "      <td>16846</td>\n",
       "      <td>CATGACCAAAGCCTAACCCTGCCACCCAGGACTCAGGCTTCCTCCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IDS-gRNA3</th>\n",
       "      <td>GAUGACCUGCGCCCCUCCCU</td>\n",
       "      <td>ENSG00000010404</td>\n",
       "      <td>IDS</td>\n",
       "      <td>GATGACCTGCGCCCCTCCCT</td>\n",
       "      <td>GATGACCTGCGCCCCTCCCT</td>\n",
       "      <td>16832</td>\n",
       "      <td>16832</td>\n",
       "      <td>CCCAGGTTGGAGGCCATGACCAAAGCCTAACCCTGCCACCCAGGAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IDS-gRNA4</th>\n",
       "      <td>CUUAUCCCCAUAACAGCCCA</td>\n",
       "      <td>ENSG00000010404</td>\n",
       "      <td>IDS</td>\n",
       "      <td>CTTATCCCCATAACAGCCCA</td>\n",
       "      <td>TGGGCTGTTATGGGGATAAG</td>\n",
       "      <td>-1</td>\n",
       "      <td>16851</td>\n",
       "      <td>CCAAAGCCTAACCCTGCCACCCAGGACTCAGGCTTCCTCCTCGAGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCDC83-gRNA1</th>\n",
       "      <td>AAAUCUGUUCUUCUUUUAAG</td>\n",
       "      <td>ENSG00000150676</td>\n",
       "      <td>CCDC83</td>\n",
       "      <td>AAATCTGTTCTTCTTTTAAG</td>\n",
       "      <td>CTTAAAAGAAGAACAGATTT</td>\n",
       "      <td>-1</td>\n",
       "      <td>27420</td>\n",
       "      <td>TTTAATAATGTCATAAGACTCTAGATCTTATTTGTAATCTTCTGTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCDC83-gRNA2</th>\n",
       "      <td>GCUUAAAAGAAGAACAGAUU</td>\n",
       "      <td>ENSG00000150676</td>\n",
       "      <td>CCDC83</td>\n",
       "      <td>GCTTAAAAGAAGAACAGATT</td>\n",
       "      <td>GCTTAAAAGAAGAACAGATT</td>\n",
       "      <td>27419</td>\n",
       "      <td>27419</td>\n",
       "      <td>TTTTAATAATGTCATAAGACTCTAGATCTTATTTGTAATCTTCTGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCDC83-gRNA3</th>\n",
       "      <td>AAGAACAGAUUUGGCACAUA</td>\n",
       "      <td>ENSG00000150676</td>\n",
       "      <td>CCDC83</td>\n",
       "      <td>AAGAACAGATTTGGCACATA</td>\n",
       "      <td>AAGAACAGATTTGGCACATA</td>\n",
       "      <td>27428</td>\n",
       "      <td>27428</td>\n",
       "      <td>TGTCATAAGACTCTAGATCTTATTTGTAATCTTCTGTTTTACCTAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCDC83-gRNA4</th>\n",
       "      <td>GCACAUACGGCAUCUACUAA</td>\n",
       "      <td>ENSG00000150676</td>\n",
       "      <td>CCDC83</td>\n",
       "      <td>GCACATACGGCATCTACTAA</td>\n",
       "      <td>GCACATACGGCATCTACTAA</td>\n",
       "      <td>27441</td>\n",
       "      <td>27441</td>\n",
       "      <td>TAGATCTTATTTGTAATCTTCTGTTTTACCTAGCTTTTACTGGCAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MAST3-gRNA1</th>\n",
       "      <td>CCCGGGAGGAUGAGCUUGAC</td>\n",
       "      <td>ENSG00000099308</td>\n",
       "      <td>MAST3</td>\n",
       "      <td>CCCGGGAGGATGAGCTTGAC</td>\n",
       "      <td>GTCAAGCTCATCCTCCCGGG</td>\n",
       "      <td>-1</td>\n",
       "      <td>25438</td>\n",
       "      <td>TCCTCAGCCTATGGGGTCAGGGCCTGACCAGAGGGGGACTTGGGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MAST3-gRNA2</th>\n",
       "      <td>ACCAGUCAAGCUCAUCCUCC</td>\n",
       "      <td>ENSG00000099308</td>\n",
       "      <td>MAST3</td>\n",
       "      <td>ACCAGTCAAGCTCATCCTCC</td>\n",
       "      <td>ACCAGTCAAGCTCATCCTCC</td>\n",
       "      <td>25434</td>\n",
       "      <td>25434</td>\n",
       "      <td>GATATCCTCAGCCTATGGGGTCAGGGCCTGACCAGAGGGGGACTTG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     gRNA sequence          GENE ID GENE NAME  \\\n",
       "IDS-gRNA1     GAACGUUCUUCUCAUCAUCG  ENSG00000010404       IDS   \n",
       "IDS-gRNA2     CCCCAUAACAGCCCAGGGAG  ENSG00000010404       IDS   \n",
       "IDS-gRNA3     GAUGACCUGCGCCCCUCCCU  ENSG00000010404       IDS   \n",
       "IDS-gRNA4     CUUAUCCCCAUAACAGCCCA  ENSG00000010404       IDS   \n",
       "CCDC83-gRNA1  AAAUCUGUUCUUCUUUUAAG  ENSG00000150676    CCDC83   \n",
       "CCDC83-gRNA2  GCUUAAAAGAAGAACAGAUU  ENSG00000150676    CCDC83   \n",
       "CCDC83-gRNA3  AAGAACAGAUUUGGCACAUA  ENSG00000150676    CCDC83   \n",
       "CCDC83-gRNA4  GCACAUACGGCAUCUACUAA  ENSG00000150676    CCDC83   \n",
       "MAST3-gRNA1   CCCGGGAGGAUGAGCUUGAC  ENSG00000099308     MAST3   \n",
       "MAST3-gRNA2   ACCAGUCAAGCUCAUCCUCC  ENSG00000099308     MAST3   \n",
       "\n",
       "                   gRNA_sequence_T gRNA_sequence_T_forward_rel_to_gene  \\\n",
       "IDS-gRNA1     GAACGTTCTTCTCATCATCG                GAACGTTCTTCTCATCATCG   \n",
       "IDS-gRNA2     CCCCATAACAGCCCAGGGAG                CTCCCTGGGCTGTTATGGGG   \n",
       "IDS-gRNA3     GATGACCTGCGCCCCTCCCT                GATGACCTGCGCCCCTCCCT   \n",
       "IDS-gRNA4     CTTATCCCCATAACAGCCCA                TGGGCTGTTATGGGGATAAG   \n",
       "CCDC83-gRNA1  AAATCTGTTCTTCTTTTAAG                CTTAAAAGAAGAACAGATTT   \n",
       "CCDC83-gRNA2  GCTTAAAAGAAGAACAGATT                GCTTAAAAGAAGAACAGATT   \n",
       "CCDC83-gRNA3  AAGAACAGATTTGGCACATA                AAGAACAGATTTGGCACATA   \n",
       "CCDC83-gRNA4  GCACATACGGCATCTACTAA                GCACATACGGCATCTACTAA   \n",
       "MAST3-gRNA1   CCCGGGAGGATGAGCTTGAC                GTCAAGCTCATCCTCCCGGG   \n",
       "MAST3-gRNA2   ACCAGTCAAGCTCATCCTCC                ACCAGTCAAGCTCATCCTCC   \n",
       "\n",
       "              gRNA reverse (-1 = True)  Forward position gRNA  \\\n",
       "IDS-gRNA1                        16810                  16810   \n",
       "IDS-gRNA2                           -1                  16846   \n",
       "IDS-gRNA3                        16832                  16832   \n",
       "IDS-gRNA4                           -1                  16851   \n",
       "CCDC83-gRNA1                        -1                  27420   \n",
       "CCDC83-gRNA2                     27419                  27419   \n",
       "CCDC83-gRNA3                     27428                  27428   \n",
       "CCDC83-gRNA4                     27441                  27441   \n",
       "MAST3-gRNA1                         -1                  25438   \n",
       "MAST3-gRNA2                      25434                  25434   \n",
       "\n",
       "                                          Seq for primer design  \n",
       "IDS-gRNA1     CACTCACTATCTCGCTTCCTCACCCAGGTTGGAGGCCATGACCAAA...  \n",
       "IDS-gRNA2     CATGACCAAAGCCTAACCCTGCCACCCAGGACTCAGGCTTCCTCCT...  \n",
       "IDS-gRNA3     CCCAGGTTGGAGGCCATGACCAAAGCCTAACCCTGCCACCCAGGAC...  \n",
       "IDS-gRNA4     CCAAAGCCTAACCCTGCCACCCAGGACTCAGGCTTCCTCCTCGAGC...  \n",
       "CCDC83-gRNA1  TTTAATAATGTCATAAGACTCTAGATCTTATTTGTAATCTTCTGTT...  \n",
       "CCDC83-gRNA2  TTTTAATAATGTCATAAGACTCTAGATCTTATTTGTAATCTTCTGT...  \n",
       "CCDC83-gRNA3  TGTCATAAGACTCTAGATCTTATTTGTAATCTTCTGTTTTACCTAG...  \n",
       "CCDC83-gRNA4  TAGATCTTATTTGTAATCTTCTGTTTTACCTAGCTTTTACTGGCAT...  \n",
       "MAST3-gRNA1   TCCTCAGCCTATGGGGTCAGGGCCTGACCAGAGGGGGACTTGGGGC...  \n",
       "MAST3-gRNA2   GATATCCTCAGCCTATGGGGTCAGGGCCTGACCAGAGGGGGACTTG...  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gRNA_predicted['Seq for primer design'] = sequences_for_primer_design\n",
    "gRNA_predicted.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "#use index to make column named guideRNA name\n",
    "gRNA_predicted['gRNA name'] = gRNA_predicted.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "primer_designed = {}\n",
    "for entry in range(len(gRNA_predicted)):\n",
    "    seq_dict = {\n",
    "        'SEQUENCE_ID': gRNA_predicted['gRNA name'].iloc[entry],\n",
    "        'SEQUENCE_TEMPLATE': gRNA_predicted['Seq for primer design'].iloc[entry],\n",
    "    }\n",
    "    primer_designed[gRNA_predicted['gRNA name'].iloc[entry]] = primer3.designPrimers(seq_dict,    \n",
    "        {\n",
    "            'PRIMER_OPT_SIZE': 20,\n",
    "            'PRIMER_PICK_INTERNAL_OLIGO': 1,\n",
    "            'PRIMER_INTERNAL_MAX_SELF_END': 8,\n",
    "            'PRIMER_MIN_SIZE': 18,\n",
    "            'PRIMER_MAX_SIZE': 25,\n",
    "            'PRIMER_OPT_TM': 60.0,\n",
    "            'PRIMER_MIN_TM': 57.0,\n",
    "            'PRIMER_MAX_TM': 63.0,\n",
    "            'PRIMER_MIN_GC': 20.0,\n",
    "            'PRIMER_MAX_GC': 80.0,\n",
    "            'PRIMER_MAX_POLY_X': 100,\n",
    "            'PRIMER_INTERNAL_MAX_POLY_X': 100,\n",
    "            'PRIMER_SALT_MONOVALENT': 50.0,\n",
    "            'PRIMER_DNA_CONC': 50.0,\n",
    "            'PRIMER_MAX_NS_ACCEPTED': 0,\n",
    "            'PRIMER_MAX_SELF_ANY': 12,\n",
    "            'PRIMER_MAX_SELF_END': 8,\n",
    "            'PRIMER_PAIR_MAX_COMPL_ANY': 12,\n",
    "            'PRIMER_PAIR_MAX_COMPL_END': 8,\n",
    "            'PRIMER_PRODUCT_SIZE_RANGE': [[375, 500]],\n",
    "        })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PRIMER_LEFT_0_SEQUENCE</th>\n",
       "      <th>PRIMER_RIGHT_0_SEQUENCE</th>\n",
       "      <th>PRIMER_PAIR_0_PRODUCT_SIZE</th>\n",
       "      <th>PRIMER_LEFT_0</th>\n",
       "      <th>PRIMER_RIGHT_0</th>\n",
       "      <th>PRIMER_INTERNAL_0</th>\n",
       "      <th>PRIMER_LEFT_0_TM</th>\n",
       "      <th>PRIMER_RIGHT_0_TM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>IDS-gRNA1</th>\n",
       "      <td>TGACCAAAGCCTAACCCTGC</td>\n",
       "      <td>CACACACCCACAGCTAGAGG</td>\n",
       "      <td>375</td>\n",
       "      <td>(38, 20)</td>\n",
       "      <td>(412, 20)</td>\n",
       "      <td>(71, 20)</td>\n",
       "      <td>60.251165</td>\n",
       "      <td>60.036952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IDS-gRNA2</th>\n",
       "      <td>TGACCAAAGCCTAACCCTGC</td>\n",
       "      <td>CACACACCCACAGCTAGAGG</td>\n",
       "      <td>375</td>\n",
       "      <td>(2, 20)</td>\n",
       "      <td>(376, 20)</td>\n",
       "      <td>(35, 20)</td>\n",
       "      <td>60.251165</td>\n",
       "      <td>60.036952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IDS-gRNA3</th>\n",
       "      <td>TGACCAAAGCCTAACCCTGC</td>\n",
       "      <td>CACACACCCACAGCTAGAGG</td>\n",
       "      <td>375</td>\n",
       "      <td>(16, 20)</td>\n",
       "      <td>(390, 20)</td>\n",
       "      <td>(49, 20)</td>\n",
       "      <td>60.251165</td>\n",
       "      <td>60.036952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IDS-gRNA4</th>\n",
       "      <td>AGGACTCAGGCTTCCTCCTC</td>\n",
       "      <td>AGATGTCCCGCACAATCTGT</td>\n",
       "      <td>424</td>\n",
       "      <td>(22, 20)</td>\n",
       "      <td>(445, 20)</td>\n",
       "      <td>(74, 20)</td>\n",
       "      <td>60.325203</td>\n",
       "      <td>59.384669</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCDC83-gRNA1</th>\n",
       "      <td>AAGTCCAGGCTTCCCACTTG</td>\n",
       "      <td>ACAGCATTTTGAACATATTGGAAGA</td>\n",
       "      <td>381</td>\n",
       "      <td>(119, 20)</td>\n",
       "      <td>(499, 25)</td>\n",
       "      <td>(143, 22)</td>\n",
       "      <td>59.888747</td>\n",
       "      <td>57.769151</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCDC83-gRNA2</th>\n",
       "      <td>TTGCTCTGCCAGGGGAAGTA</td>\n",
       "      <td>TGGCACAACTCTATGATAGTTTTCT</td>\n",
       "      <td>379</td>\n",
       "      <td>(67, 20)</td>\n",
       "      <td>(445, 25)</td>\n",
       "      <td>(121, 21)</td>\n",
       "      <td>60.840288</td>\n",
       "      <td>58.288723</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCDC83-gRNA3</th>\n",
       "      <td>AAGTCCAGGCTTCCCACTTG</td>\n",
       "      <td>ACAGCATTTTGAACATATTGGAAGA</td>\n",
       "      <td>381</td>\n",
       "      <td>(111, 20)</td>\n",
       "      <td>(491, 25)</td>\n",
       "      <td>(135, 22)</td>\n",
       "      <td>59.888747</td>\n",
       "      <td>57.769151</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCDC83-gRNA4</th>\n",
       "      <td>CCCACTTGCCTTCCTGTGAT</td>\n",
       "      <td>GCCGTTTAAAATACACTTAACAGCA</td>\n",
       "      <td>388</td>\n",
       "      <td>(110, 20)</td>\n",
       "      <td>(497, 25)</td>\n",
       "      <td>(130, 21)</td>\n",
       "      <td>59.961256</td>\n",
       "      <td>58.855825</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MAST3-gRNA1</th>\n",
       "      <td>TGCATAGGAGCTAGGCAGGA</td>\n",
       "      <td>GTGAAGGAGGAGCCACAACT</td>\n",
       "      <td>375</td>\n",
       "      <td>(112, 20)</td>\n",
       "      <td>(486, 20)</td>\n",
       "      <td>(194, 20)</td>\n",
       "      <td>60.105504</td>\n",
       "      <td>59.601682</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MAST3-gRNA2</th>\n",
       "      <td>TGCATAGGAGCTAGGCAGGA</td>\n",
       "      <td>GTGAAGGAGGAGCCACAACT</td>\n",
       "      <td>375</td>\n",
       "      <td>(116, 20)</td>\n",
       "      <td>(490, 20)</td>\n",
       "      <td>(198, 20)</td>\n",
       "      <td>60.105504</td>\n",
       "      <td>59.601682</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             PRIMER_LEFT_0_SEQUENCE    PRIMER_RIGHT_0_SEQUENCE  \\\n",
       "IDS-gRNA1      TGACCAAAGCCTAACCCTGC       CACACACCCACAGCTAGAGG   \n",
       "IDS-gRNA2      TGACCAAAGCCTAACCCTGC       CACACACCCACAGCTAGAGG   \n",
       "IDS-gRNA3      TGACCAAAGCCTAACCCTGC       CACACACCCACAGCTAGAGG   \n",
       "IDS-gRNA4      AGGACTCAGGCTTCCTCCTC       AGATGTCCCGCACAATCTGT   \n",
       "CCDC83-gRNA1   AAGTCCAGGCTTCCCACTTG  ACAGCATTTTGAACATATTGGAAGA   \n",
       "CCDC83-gRNA2   TTGCTCTGCCAGGGGAAGTA  TGGCACAACTCTATGATAGTTTTCT   \n",
       "CCDC83-gRNA3   AAGTCCAGGCTTCCCACTTG  ACAGCATTTTGAACATATTGGAAGA   \n",
       "CCDC83-gRNA4   CCCACTTGCCTTCCTGTGAT  GCCGTTTAAAATACACTTAACAGCA   \n",
       "MAST3-gRNA1    TGCATAGGAGCTAGGCAGGA       GTGAAGGAGGAGCCACAACT   \n",
       "MAST3-gRNA2    TGCATAGGAGCTAGGCAGGA       GTGAAGGAGGAGCCACAACT   \n",
       "\n",
       "              PRIMER_PAIR_0_PRODUCT_SIZE PRIMER_LEFT_0 PRIMER_RIGHT_0  \\\n",
       "IDS-gRNA1                            375      (38, 20)      (412, 20)   \n",
       "IDS-gRNA2                            375       (2, 20)      (376, 20)   \n",
       "IDS-gRNA3                            375      (16, 20)      (390, 20)   \n",
       "IDS-gRNA4                            424      (22, 20)      (445, 20)   \n",
       "CCDC83-gRNA1                         381     (119, 20)      (499, 25)   \n",
       "CCDC83-gRNA2                         379      (67, 20)      (445, 25)   \n",
       "CCDC83-gRNA3                         381     (111, 20)      (491, 25)   \n",
       "CCDC83-gRNA4                         388     (110, 20)      (497, 25)   \n",
       "MAST3-gRNA1                          375     (112, 20)      (486, 20)   \n",
       "MAST3-gRNA2                          375     (116, 20)      (490, 20)   \n",
       "\n",
       "             PRIMER_INTERNAL_0  PRIMER_LEFT_0_TM  PRIMER_RIGHT_0_TM  \n",
       "IDS-gRNA1             (71, 20)         60.251165          60.036952  \n",
       "IDS-gRNA2             (35, 20)         60.251165          60.036952  \n",
       "IDS-gRNA3             (49, 20)         60.251165          60.036952  \n",
       "IDS-gRNA4             (74, 20)         60.325203          59.384669  \n",
       "CCDC83-gRNA1         (143, 22)         59.888747          57.769151  \n",
       "CCDC83-gRNA2         (121, 21)         60.840288          58.288723  \n",
       "CCDC83-gRNA3         (135, 22)         59.888747          57.769151  \n",
       "CCDC83-gRNA4         (130, 21)         59.961256          58.855825  \n",
       "MAST3-gRNA1          (194, 20)         60.105504          59.601682  \n",
       "MAST3-gRNA2          (198, 20)         60.105504          59.601682  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "primer_df = pd.DataFrame.from_dict(primer_designed, orient='index')\n",
    "primer_df = primer_df[[\"PRIMER_LEFT_0_SEQUENCE\", \"PRIMER_RIGHT_0_SEQUENCE\", \"PRIMER_PAIR_0_PRODUCT_SIZE\", \"PRIMER_LEFT_0\", \"PRIMER_RIGHT_0\", \"PRIMER_INTERNAL_0\", \"PRIMER_LEFT_0_TM\", \"PRIMER_RIGHT_0_TM\"]]\n",
    "primer_df.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gRNA sequence</th>\n",
       "      <th>GENE ID</th>\n",
       "      <th>GENE NAME</th>\n",
       "      <th>gRNA_sequence_T</th>\n",
       "      <th>gRNA_sequence_T_forward_rel_to_gene</th>\n",
       "      <th>Forward position gRNA</th>\n",
       "      <th>PRIMER_LEFT_0_SEQUENCE</th>\n",
       "      <th>PRIMER_RIGHT_0_SEQUENCE</th>\n",
       "      <th>PRIMER_PAIR_0_PRODUCT_SIZE</th>\n",
       "      <th>PRIMER_LEFT_0</th>\n",
       "      <th>PRIMER_RIGHT_0</th>\n",
       "      <th>PRIMER_INTERNAL_0</th>\n",
       "      <th>PRIMER_LEFT_0_TM</th>\n",
       "      <th>PRIMER_RIGHT_0_TM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>IDS-gRNA1</th>\n",
       "      <td>GAACGUUCUUCUCAUCAUCG</td>\n",
       "      <td>ENSG00000010404</td>\n",
       "      <td>IDS</td>\n",
       "      <td>GAACGTTCTTCTCATCATCG</td>\n",
       "      <td>GAACGTTCTTCTCATCATCG</td>\n",
       "      <td>16810</td>\n",
       "      <td>TGACCAAAGCCTAACCCTGC</td>\n",
       "      <td>CACACACCCACAGCTAGAGG</td>\n",
       "      <td>375</td>\n",
       "      <td>(38, 20)</td>\n",
       "      <td>(412, 20)</td>\n",
       "      <td>(71, 20)</td>\n",
       "      <td>60.251165</td>\n",
       "      <td>60.036952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IDS-gRNA2</th>\n",
       "      <td>CCCCAUAACAGCCCAGGGAG</td>\n",
       "      <td>ENSG00000010404</td>\n",
       "      <td>IDS</td>\n",
       "      <td>CCCCATAACAGCCCAGGGAG</td>\n",
       "      <td>CTCCCTGGGCTGTTATGGGG</td>\n",
       "      <td>16846</td>\n",
       "      <td>TGACCAAAGCCTAACCCTGC</td>\n",
       "      <td>CACACACCCACAGCTAGAGG</td>\n",
       "      <td>375</td>\n",
       "      <td>(2, 20)</td>\n",
       "      <td>(376, 20)</td>\n",
       "      <td>(35, 20)</td>\n",
       "      <td>60.251165</td>\n",
       "      <td>60.036952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IDS-gRNA3</th>\n",
       "      <td>GAUGACCUGCGCCCCUCCCU</td>\n",
       "      <td>ENSG00000010404</td>\n",
       "      <td>IDS</td>\n",
       "      <td>GATGACCTGCGCCCCTCCCT</td>\n",
       "      <td>GATGACCTGCGCCCCTCCCT</td>\n",
       "      <td>16832</td>\n",
       "      <td>TGACCAAAGCCTAACCCTGC</td>\n",
       "      <td>CACACACCCACAGCTAGAGG</td>\n",
       "      <td>375</td>\n",
       "      <td>(16, 20)</td>\n",
       "      <td>(390, 20)</td>\n",
       "      <td>(49, 20)</td>\n",
       "      <td>60.251165</td>\n",
       "      <td>60.036952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IDS-gRNA4</th>\n",
       "      <td>CUUAUCCCCAUAACAGCCCA</td>\n",
       "      <td>ENSG00000010404</td>\n",
       "      <td>IDS</td>\n",
       "      <td>CTTATCCCCATAACAGCCCA</td>\n",
       "      <td>TGGGCTGTTATGGGGATAAG</td>\n",
       "      <td>16851</td>\n",
       "      <td>AGGACTCAGGCTTCCTCCTC</td>\n",
       "      <td>AGATGTCCCGCACAATCTGT</td>\n",
       "      <td>424</td>\n",
       "      <td>(22, 20)</td>\n",
       "      <td>(445, 20)</td>\n",
       "      <td>(74, 20)</td>\n",
       "      <td>60.325203</td>\n",
       "      <td>59.384669</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCDC83-gRNA1</th>\n",
       "      <td>AAAUCUGUUCUUCUUUUAAG</td>\n",
       "      <td>ENSG00000150676</td>\n",
       "      <td>CCDC83</td>\n",
       "      <td>AAATCTGTTCTTCTTTTAAG</td>\n",
       "      <td>CTTAAAAGAAGAACAGATTT</td>\n",
       "      <td>27420</td>\n",
       "      <td>AAGTCCAGGCTTCCCACTTG</td>\n",
       "      <td>ACAGCATTTTGAACATATTGGAAGA</td>\n",
       "      <td>381</td>\n",
       "      <td>(119, 20)</td>\n",
       "      <td>(499, 25)</td>\n",
       "      <td>(143, 22)</td>\n",
       "      <td>59.888747</td>\n",
       "      <td>57.769151</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCDC83-gRNA2</th>\n",
       "      <td>GCUUAAAAGAAGAACAGAUU</td>\n",
       "      <td>ENSG00000150676</td>\n",
       "      <td>CCDC83</td>\n",
       "      <td>GCTTAAAAGAAGAACAGATT</td>\n",
       "      <td>GCTTAAAAGAAGAACAGATT</td>\n",
       "      <td>27419</td>\n",
       "      <td>TTGCTCTGCCAGGGGAAGTA</td>\n",
       "      <td>TGGCACAACTCTATGATAGTTTTCT</td>\n",
       "      <td>379</td>\n",
       "      <td>(67, 20)</td>\n",
       "      <td>(445, 25)</td>\n",
       "      <td>(121, 21)</td>\n",
       "      <td>60.840288</td>\n",
       "      <td>58.288723</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCDC83-gRNA3</th>\n",
       "      <td>AAGAACAGAUUUGGCACAUA</td>\n",
       "      <td>ENSG00000150676</td>\n",
       "      <td>CCDC83</td>\n",
       "      <td>AAGAACAGATTTGGCACATA</td>\n",
       "      <td>AAGAACAGATTTGGCACATA</td>\n",
       "      <td>27428</td>\n",
       "      <td>AAGTCCAGGCTTCCCACTTG</td>\n",
       "      <td>ACAGCATTTTGAACATATTGGAAGA</td>\n",
       "      <td>381</td>\n",
       "      <td>(111, 20)</td>\n",
       "      <td>(491, 25)</td>\n",
       "      <td>(135, 22)</td>\n",
       "      <td>59.888747</td>\n",
       "      <td>57.769151</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCDC83-gRNA4</th>\n",
       "      <td>GCACAUACGGCAUCUACUAA</td>\n",
       "      <td>ENSG00000150676</td>\n",
       "      <td>CCDC83</td>\n",
       "      <td>GCACATACGGCATCTACTAA</td>\n",
       "      <td>GCACATACGGCATCTACTAA</td>\n",
       "      <td>27441</td>\n",
       "      <td>CCCACTTGCCTTCCTGTGAT</td>\n",
       "      <td>GCCGTTTAAAATACACTTAACAGCA</td>\n",
       "      <td>388</td>\n",
       "      <td>(110, 20)</td>\n",
       "      <td>(497, 25)</td>\n",
       "      <td>(130, 21)</td>\n",
       "      <td>59.961256</td>\n",
       "      <td>58.855825</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MAST3-gRNA1</th>\n",
       "      <td>CCCGGGAGGAUGAGCUUGAC</td>\n",
       "      <td>ENSG00000099308</td>\n",
       "      <td>MAST3</td>\n",
       "      <td>CCCGGGAGGATGAGCTTGAC</td>\n",
       "      <td>GTCAAGCTCATCCTCCCGGG</td>\n",
       "      <td>25438</td>\n",
       "      <td>TGCATAGGAGCTAGGCAGGA</td>\n",
       "      <td>GTGAAGGAGGAGCCACAACT</td>\n",
       "      <td>375</td>\n",
       "      <td>(112, 20)</td>\n",
       "      <td>(486, 20)</td>\n",
       "      <td>(194, 20)</td>\n",
       "      <td>60.105504</td>\n",
       "      <td>59.601682</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MAST3-gRNA2</th>\n",
       "      <td>ACCAGUCAAGCUCAUCCUCC</td>\n",
       "      <td>ENSG00000099308</td>\n",
       "      <td>MAST3</td>\n",
       "      <td>ACCAGTCAAGCTCATCCTCC</td>\n",
       "      <td>ACCAGTCAAGCTCATCCTCC</td>\n",
       "      <td>25434</td>\n",
       "      <td>TGCATAGGAGCTAGGCAGGA</td>\n",
       "      <td>GTGAAGGAGGAGCCACAACT</td>\n",
       "      <td>375</td>\n",
       "      <td>(116, 20)</td>\n",
       "      <td>(490, 20)</td>\n",
       "      <td>(198, 20)</td>\n",
       "      <td>60.105504</td>\n",
       "      <td>59.601682</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     gRNA sequence          GENE ID GENE NAME  \\\n",
       "IDS-gRNA1     GAACGUUCUUCUCAUCAUCG  ENSG00000010404       IDS   \n",
       "IDS-gRNA2     CCCCAUAACAGCCCAGGGAG  ENSG00000010404       IDS   \n",
       "IDS-gRNA3     GAUGACCUGCGCCCCUCCCU  ENSG00000010404       IDS   \n",
       "IDS-gRNA4     CUUAUCCCCAUAACAGCCCA  ENSG00000010404       IDS   \n",
       "CCDC83-gRNA1  AAAUCUGUUCUUCUUUUAAG  ENSG00000150676    CCDC83   \n",
       "CCDC83-gRNA2  GCUUAAAAGAAGAACAGAUU  ENSG00000150676    CCDC83   \n",
       "CCDC83-gRNA3  AAGAACAGAUUUGGCACAUA  ENSG00000150676    CCDC83   \n",
       "CCDC83-gRNA4  GCACAUACGGCAUCUACUAA  ENSG00000150676    CCDC83   \n",
       "MAST3-gRNA1   CCCGGGAGGAUGAGCUUGAC  ENSG00000099308     MAST3   \n",
       "MAST3-gRNA2   ACCAGUCAAGCUCAUCCUCC  ENSG00000099308     MAST3   \n",
       "\n",
       "                   gRNA_sequence_T gRNA_sequence_T_forward_rel_to_gene  \\\n",
       "IDS-gRNA1     GAACGTTCTTCTCATCATCG                GAACGTTCTTCTCATCATCG   \n",
       "IDS-gRNA2     CCCCATAACAGCCCAGGGAG                CTCCCTGGGCTGTTATGGGG   \n",
       "IDS-gRNA3     GATGACCTGCGCCCCTCCCT                GATGACCTGCGCCCCTCCCT   \n",
       "IDS-gRNA4     CTTATCCCCATAACAGCCCA                TGGGCTGTTATGGGGATAAG   \n",
       "CCDC83-gRNA1  AAATCTGTTCTTCTTTTAAG                CTTAAAAGAAGAACAGATTT   \n",
       "CCDC83-gRNA2  GCTTAAAAGAAGAACAGATT                GCTTAAAAGAAGAACAGATT   \n",
       "CCDC83-gRNA3  AAGAACAGATTTGGCACATA                AAGAACAGATTTGGCACATA   \n",
       "CCDC83-gRNA4  GCACATACGGCATCTACTAA                GCACATACGGCATCTACTAA   \n",
       "MAST3-gRNA1   CCCGGGAGGATGAGCTTGAC                GTCAAGCTCATCCTCCCGGG   \n",
       "MAST3-gRNA2   ACCAGTCAAGCTCATCCTCC                ACCAGTCAAGCTCATCCTCC   \n",
       "\n",
       "              Forward position gRNA PRIMER_LEFT_0_SEQUENCE  \\\n",
       "IDS-gRNA1                     16810   TGACCAAAGCCTAACCCTGC   \n",
       "IDS-gRNA2                     16846   TGACCAAAGCCTAACCCTGC   \n",
       "IDS-gRNA3                     16832   TGACCAAAGCCTAACCCTGC   \n",
       "IDS-gRNA4                     16851   AGGACTCAGGCTTCCTCCTC   \n",
       "CCDC83-gRNA1                  27420   AAGTCCAGGCTTCCCACTTG   \n",
       "CCDC83-gRNA2                  27419   TTGCTCTGCCAGGGGAAGTA   \n",
       "CCDC83-gRNA3                  27428   AAGTCCAGGCTTCCCACTTG   \n",
       "CCDC83-gRNA4                  27441   CCCACTTGCCTTCCTGTGAT   \n",
       "MAST3-gRNA1                   25438   TGCATAGGAGCTAGGCAGGA   \n",
       "MAST3-gRNA2                   25434   TGCATAGGAGCTAGGCAGGA   \n",
       "\n",
       "                PRIMER_RIGHT_0_SEQUENCE  PRIMER_PAIR_0_PRODUCT_SIZE  \\\n",
       "IDS-gRNA1          CACACACCCACAGCTAGAGG                         375   \n",
       "IDS-gRNA2          CACACACCCACAGCTAGAGG                         375   \n",
       "IDS-gRNA3          CACACACCCACAGCTAGAGG                         375   \n",
       "IDS-gRNA4          AGATGTCCCGCACAATCTGT                         424   \n",
       "CCDC83-gRNA1  ACAGCATTTTGAACATATTGGAAGA                         381   \n",
       "CCDC83-gRNA2  TGGCACAACTCTATGATAGTTTTCT                         379   \n",
       "CCDC83-gRNA3  ACAGCATTTTGAACATATTGGAAGA                         381   \n",
       "CCDC83-gRNA4  GCCGTTTAAAATACACTTAACAGCA                         388   \n",
       "MAST3-gRNA1        GTGAAGGAGGAGCCACAACT                         375   \n",
       "MAST3-gRNA2        GTGAAGGAGGAGCCACAACT                         375   \n",
       "\n",
       "             PRIMER_LEFT_0 PRIMER_RIGHT_0 PRIMER_INTERNAL_0  PRIMER_LEFT_0_TM  \\\n",
       "IDS-gRNA1         (38, 20)      (412, 20)          (71, 20)         60.251165   \n",
       "IDS-gRNA2          (2, 20)      (376, 20)          (35, 20)         60.251165   \n",
       "IDS-gRNA3         (16, 20)      (390, 20)          (49, 20)         60.251165   \n",
       "IDS-gRNA4         (22, 20)      (445, 20)          (74, 20)         60.325203   \n",
       "CCDC83-gRNA1     (119, 20)      (499, 25)         (143, 22)         59.888747   \n",
       "CCDC83-gRNA2      (67, 20)      (445, 25)         (121, 21)         60.840288   \n",
       "CCDC83-gRNA3     (111, 20)      (491, 25)         (135, 22)         59.888747   \n",
       "CCDC83-gRNA4     (110, 20)      (497, 25)         (130, 21)         59.961256   \n",
       "MAST3-gRNA1      (112, 20)      (486, 20)         (194, 20)         60.105504   \n",
       "MAST3-gRNA2      (116, 20)      (490, 20)         (198, 20)         60.105504   \n",
       "\n",
       "              PRIMER_RIGHT_0_TM  \n",
       "IDS-gRNA1             60.036952  \n",
       "IDS-gRNA2             60.036952  \n",
       "IDS-gRNA3             60.036952  \n",
       "IDS-gRNA4             59.384669  \n",
       "CCDC83-gRNA1          57.769151  \n",
       "CCDC83-gRNA2          58.288723  \n",
       "CCDC83-gRNA3          57.769151  \n",
       "CCDC83-gRNA4          58.855825  \n",
       "MAST3-gRNA1           59.601682  \n",
       "MAST3-gRNA2           59.601682  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gRNA_primer_df = gRNA_predicted.join(primer_df)\n",
    "gRNA_primer_df = gRNA_primer_df.drop(columns=[\"gRNA name\", \"gRNA reverse (-1 = True)\"])\n",
    "gRNA_primer_df_wo_seq = gRNA_primer_df.drop(columns=[\"Seq for primer design\"])\n",
    "gRNA_primer_df_wo_seq.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "gRNA_primer_df_wo_seq.to_csv(project_id + '_gRNA_primer.csv')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
